intercom-common/src/tyhandlers.rs

```diff
@@ -93,80 +104,91 @@ impl TypeHandler for ComItfParam {
 }
 
 /// String parameter handler. Converts between Rust String and COM BSTR types.
-struct StringParam( Type );
+struct StringParam { ty: Type, context: TypeContext }
 impl TypeHandler for StringParam
 {
-    fn rust_ty( &self ) -> Type { self.0.clone() }
+    fn rust_ty( &self ) -> Type { self.ty.clone() }
 
     fn com_ty( &self ) -> Type
     {
-        parse_quote!( ::intercom::BStr )
-    }
-
-    fn com_to_rust( &self, ident : &Ident ) -> ComToRust
-    {
-        ComToRust {
-            stack: None,
-            conversion: quote!( #ident.into() )
+        match self.context.dir {
+            Direction::In => parse_quote!( ::intercom::raw::InBSTR ),
+            Direction::Out | Direction::Retval => parse_quote!( ::intercom::raw::OutBSTR ),
         }
     }
 
-    fn rust_to_com( &self, ident : &Ident ) -> Tokens
+    fn com_to_rust( &self, ident : Ident ) -> Tokens
     {
-        quote!( #ident.into() )
-    }
-}
+        match self.context.dir {
 
-/// String parameter handler. Converts between Rust &str and COM BSTR types.
-struct StringRefParam( Type );
-impl TypeHandler for StringRefParam
-{
-    fn rust_ty( &self ) -> Type { self.0.clone() }
+            Direction::In => {
 
-    fn com_ty( &self ) -> Type
-    {
-        parse_quote!( ::intercom::BStr )
+                let rust_ty = self.rust_ty();
+                quote!(
+                    < #rust_ty as ::intercom::FromWithTemporary<&::intercom::BStr> >::from_temporary(
+                        &mut < #rust_ty as ::intercom::FromWithTemporary<&::intercom::BStr> >::to_temporary(
+                            ::intercom::BStr::from_ptr( #ident ) )? )?
+                )
```

Comment:
How many copies of the data this will actually make?

Reply:
Shouldn't make any unnecessary ones. The `to_temporary` will move the value. The `from_temporary` will create a copy.

If there is no need for a temporary, the `to_temporary` is able to pass the original item through without copying. For the cases where the callee needs an owned value, the `from_temporary` needs to create a copy anyway as COM can't really pass ownership of `IN` strings. In places where the callee needs a reference such as `&BStr`, the `from_temporary` might turn a `&BStr` into a copy, but that copy is a pointer copy instead of data copy so more or less normal parameter passing.